### PR TITLE
refactor: refactor device discovery process

### PR DESCRIPTION
**Pull request by David-T-White**
Change logic in discoverDevices to increase chances of discovering multipath devices properly by retrying discovery multiple times if only non-multipath device is found. Refactor FindDiskById (and make it public) to not read the entire disk/by-id directory during device search

### CODE REVIEW

File: sas/sas.go [hunk 1]
```diff
@@ -201,15 +202,17 @@ func discoverDevices(logger klog.Logger, c *Connector, io ioHandler) error {
...
 	// two-phase search:
 	// first phase, search existing device paths, if a multipath dm is found, exit loop
 	// otherwise, in second phase, rescan scsi bus and search again, return with any findings
-	for true {
+	for retries < 5 {
 
 		// Find the multipath device using WWN
-		dm, devices = findDiskById(logger, c.VolumeWWN, io)
+		logger.V(5).Info("Sleeping 0.5 seconds before device discovery")
+		time.Sleep(500 * time.Millisecond)
```
Comment (seagate-chris) [inline sas/sas.go:214]: It'd be good to drop a comment explaining how you came up with this value, e.g. is it based on some concrete detail or is it just something that worked on your system and might need to be larger for a different configuration?

**Review by seagate-chris — APPROVED** [review]
Looks good!

File: sas/sas.go [hunk 2]
```diff
@@ -201,15 +202,17 @@ func discoverDevices(logger klog.Logger, c *Connector, io ioHandler) error {
...
 	c.Multipath = false
 	c.OSPathName = ""
-	rescaned := false
+	retries := 0
 
 	// two-phase search:
 	// first phase, search existing device paths, if a multipath dm is found, exit loop
 	// otherwise, in second phase, rescan scsi bus and search again, return with any findings
-	for true {
+	for retries < 5 {
```
Comment (jskazinski) [inline sas/sas.go:210]: Suggest we make '5' a const value

File: sas/sas.go (hunk 1 above)
Comment (jskazinski) [reply sas/sas.go:214]: Also suggest adding the Sleep time as a const (with a comment)

**Review by jskazinski — APPROVED** [review]
Looks good, but would prefer to see the use of const values

**Pull request merged**